packages/flutter/lib/src/widgets/overscroll_indicator.dart

```diff
@@ -834,142 +829,177 @@ class _StretchingOverscrollIndicatorState extends State<StretchingOverscrollIndi
-    _stretchSizeTween.end = linearIntensity + exponentialIntensity;
-    _stretchController.duration = _stretchDuration;
-    if (_state != _StretchState.pull) {
-      _stretchController.forward(from: 0.0);
-      _state = _StretchState.pull;
-    } else {
-      if (!_stretchController.isAnimating) {
-        assert(_stretchController.value == 1.0);
-        notifyListeners();
-      }
-    }
+  /// Creates a spring description with the given mass,
+  /// using predefined stiffness and damping values for the spring.
+  SpringDescription createSpringDescription(double mass) {
+    return SpringDescription(mass: mass, stiffness: _springStiffness, damping: _springDamping);
   }
 
-  void scrollEnd() {
-    if (_state == _StretchState.pull) {
-      _recede(_stretchDuration);
+  /// Since the implementation methods of Android and Flutter differ,
+  /// a [SpringSimulation] instance is created using an estimated
+  /// value nearly identical to the Spring animation of the Stretch Overscroll effect.
+  SpringSimulation createSimulation(double velocity) {
+    return SpringSimulation(createSpringDescription(1.0), overscroll, 0.0, velocity);
+  }
+
+  /// Uses the saturation function approach to scale velocity
+  /// Small velocities remain roughly proportional, while large
+  /// velocities gradually saturate. Returns the scaled velocity
+  /// with the original direction preserved.
+  double _calculateVelocityScale(double velocity) {
+    const double referenceVelocity = 3000.0;
+    const double velocityScaleFactor = 0.0001;
+    final double absVelocity = velocity.abs();
+
+    // Normalizes using 3,000px/s as the reference.
+    final double normalizedVelocity = absVelocity / referenceVelocity;
+    final double saturatedScale = normalizedVelocity / (1.0 + normalizedVelocity);
+    final double maxScale = 1.0 + velocity.abs() * velocityScaleFactor;
+
+    return saturatedScale * maxScale * velocity.sign;
+  }
+
+  /// Handle a fling to the edge of the viewport at a particular velocity.
+  ///
+  /// The velocity must be positive.
+  void absorbImpact(double velocity) {
+    if (velocity == 0.0) {
+      return;
     }
+    final double scaledVelocity = _calculateVelocityScale(velocity);
+    animate(SpringSimulation(createSpringDescription(0.5), overscroll, 0.0, scaledVelocity));
   }
 
-  void _changePhase(AnimationStatus status) {
-    if (!status.isCompleted) {
+  /// Called when the overscroll ends to trigger a fling animation if needed.
+  void scrollEnd(double velocity) {
+    if (velocity == 0.0 && overscroll == 0.0) {
       return;
     }
-    switch (_state) {
-      case _StretchState.absorb:
-        _recede(_stretchDuration);
-      case _StretchState.recede:
-        _state = _StretchState.idle;
-        _pullDistance = 0.0;
-      case _StretchState.pull:
-      case _StretchState.idle:
-        break;
+    final double scaledVelocity = -(velocity * _flingVelocityFriction);
```

Comment:
I'd imagine that the starting velocity to be _equal_ to the fling velocity. The velocity scaling is probably due to different unit system. If I remember correctly, `SpringSimulation` takes the full range (the distance from the starting point to the ending point) as 1.0, while `DragDetails` uses the number of pixels. Therefore the conversion rate is actually the height of the widget in pixels.